developer_manual/client_apis/OCS/ocs-openapi.rst

```diff
@@ -0,0 +1,731 @@
+
+The closer you narrow down a type without violating any constraints the better the resulting specification will be.
+Psalm will catch these problems for you if you configured the issue handlers mentioned above correctly.
+
+.. collapse:: Examples
+
+    .. code-block:: php
+        :caption: Bad
+
+        /**
+         * @return array
+         */
+        public function someHelperMethod() {
+            ...
+            return [
+                "id" => id,
+                "name" => name,
+            ];
+        }
+
+    .. code-block:: php
+        :caption: Good
+
+        /**
+         * @return array{id: int, name: string}
+         */
+        public function someHelperMethod() {
+            ...
+            return [
+                "id" => id,
+                "name" => name,
+            ];
+        }
+
+PREFER to use `null` to represent empty data
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Your API should be designed in a way that represents empty data with `null`.
+
+There is a problem with PHP and arrays that get converted to JSON.
+JSON has lists and objects while PHP only has arrays.
+If you were to return an empty array in PHP it will always turn into `[]` in JSON.
+This is not a problem for endpoints that always return lists, but most endpoints return a single JSON object.
+For those endpoints returning `[]` in PHP is a problem because the consumer will either get `[]` or `{...}` which is hard to handle.
+
+If you are not able to use `null` for whatever reason, use `new \stdClass()` instead.
+It will get correctly converted into `{}` in the JSON response.
+
+If you are working with an existing API where you can not break compatibility, you can also type the result as `array<empty>`.
+
+.. collapse:: Examples
+
+    .. code-block:: php
+        :caption: Bad
+
+        /**
+         * @return DataResponse<Http::STATUS_OK, array, array{}>
+         */
+        public function someControllerMethod() {
+            ...
+            return DataResponse([]);
+        }
+
+    .. code-block:: php
+        :caption: Good
+
+        /**
+         * @return DataResponse<Http::STATUS_OK, null, array{}>
+         */
+        public function someControllerMethod() {
+            ...
+            return DataResponse(null);
+        }
+
+        /**
+         * @return DataResponse<Http::STATUS_OK, \stdClass, array{}>
+         */
+        public function someControllerMethod() {
+            ...
+            return DataResponse(new \stdClass());
+        }
+
+        /**
+         * @return DataResponse<Http::STATUS_OK, array<empty>, array{}>
+         */
+        public function someControllerMethod() {
+            ...
+            return DataResponse([]);
+        }
+
+DO NOT throw exceptions
+^^^^^^^^^^^^^^^^^^^^^^^
+
+Return valid responses with an error message instead.
+
+.. collapse:: Examples
+
+    .. code-block:: php
+        :caption: Bad
+
+        /**
+         * @throws OCSBadRequestException
```

Comment:
Those `OCS…Exception`s are OCP public API. If usage is discouraged, it should be marked deprecated with explanation and the "Good" be included in the deprecation.

Reply:
This is only advice for writing good APIs. You can still use those exceptions and there are valid use cases (especially for keeping API compatibility).